# EDA

## Imports

In [1]:
import pandas as pd

## Cargar data

In [2]:
ROUTE_STORE = r"../data/raw/store.csv"
ROUTE_TRAIN = r"../data/raw/train.csv"

In [3]:
df_store = pd.read_csv(ROUTE_STORE)
df_train = pd.read_csv(ROUTE_TRAIN)

C:\Users\Carlos\AppData\Local\Temp\ipykernel_26592\4020586428.py:2: DtypeWarning: Columns (0: StateHoliday) have mixed types. Specify dtype option on import or set low_memory=False.
  df_train = pd.read_csv(ROUTE_TRAIN)


## Entendimiento de los datos

### Store

In [4]:
df_store.head()

,Store,StoreType,Assortment,CompetitionDistance,CompetitionOpenSinceMonth,CompetitionOpenSinceYear,Promo2,Promo2SinceWeek,Promo2SinceYear,PromoInterval
0,1,c,a,1270.0,9.0,2008.0,0,NaN,NaN,NaN
1,2,a,a,570.0,11.0,2007.0,1,13.0,2010.0,"Jan,Apr,Jul,Oct"
2,3,a,a,14130.0,12.0,2006.0,1,14.0,2011.0,"Jan,Apr,Jul,Oct"
3,4,c,c,620.0,9.0,2009.0,0,NaN,NaN,NaN
4,5,a,a,29910.0,4.0,2015.0,0,NaN,NaN,NaN


In [5]:
df_store.info()

<class 'pandas.DataFrame'>
RangeIndex: 1115 entries, 0 to 1114
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Store                      1115 non-null   int64  
 1   StoreType                  1115 non-null   str    
 2   Assortment                 1115 non-null   str    
 3   CompetitionDistance        1112 non-null   float64
 4   CompetitionOpenSinceMonth  761 non-null    float64
 5   CompetitionOpenSinceYear   761 non-null    float64
 6   Promo2                     1115 non-null   int64  
 7   Promo2SinceWeek            571 non-null    float64
 8   Promo2SinceYear            571 non-null    float64
 9   PromoInterval              571 non-null    str    
dtypes: float64(5), int64(2), str(3)
memory usage: 87.2 KB


In [6]:
df_store.describe()

,Store,CompetitionDistance,CompetitionOpenSinceMonth,CompetitionOpenSinceYear,Promo2,Promo2SinceWeek,Promo2SinceYear
count,1115.00000,1112.000000,761.000000,761.000000,1115.000000,571.000000,571.000000
mean,558.00000,5404.901079,7.224704,2008.668857,0.512108,23.595447,2011.763573
std,322.01708,7663.174720,3.212348,6.195983,0.500078,14.141984,1.674935
min,1.00000,20.000000,1.000000,1900.000000,0.000000,1.000000,2009.000000
25%,279.50000,717.500000,4.000000,2006.000000,0.000000,13.000000,2011.000000
50%,558.00000,2325.000000,8.000000,2010.000000,1.000000,22.000000,2012.000000
75%,836.50000,6882.500000,10.000000,2013.000000,1.000000,37.000000,2013.000000
max,1115.00000,75860.000000,12.000000,2015.000000,1.000000,50.000000,2015.000000


### Train

In [7]:
df_train.head()

,Store,DayOfWeek,Date,Sales,Customers,Open,Promo,StateHoliday,SchoolHoliday
0,1,5,2015-07-31,5263,555,1,1,0,1
1,2,5,2015-07-31,6064,625,1,1,0,1
2,3,5,2015-07-31,8314,821,1,1,0,1
3,4,5,2015-07-31,13995,1498,1,1,0,1
4,5,5,2015-07-31,4822,559,1,1,0,1


In [8]:
df_train.info()

<class 'pandas.DataFrame'>
RangeIndex: 1017209 entries, 0 to 1017208
Data columns (total 9 columns):
 #   Column         Non-Null Count    Dtype 
---  ------         --------------    ----- 
 0   Store          1017209 non-null  int64 
 1   DayOfWeek      1017209 non-null  int64 
 2   Date           1017209 non-null  str   
 3   Sales          1017209 non-null  int64 
 4   Customers      1017209 non-null  int64 
 5   Open           1017209 non-null  int64 
 6   Promo          1017209 non-null  int64 
 7   StateHoliday   1017209 non-null  object
 8   SchoolHoliday  1017209 non-null  int64 
dtypes: int64(7), object(1), str(1)
memory usage: 69.8+ MB


In [9]:
df_train.describe()

,Store,DayOfWeek,Sales,Customers,Open,Promo,SchoolHoliday
count,1.017209e+06,1.017209e+06,1.017209e+06,1.017209e+06,1.017209e+06,1.017209e+06,1.017209e+06
mean,5.584297e+02,3.998341e+00,5.773819e+03,6.331459e+02,8.301067e-01,3.815145e-01,1.786467e-01
std,3.219087e+02,1.997391e+00,3.849926e+03,4.644117e+02,3.755392e-01,4.857586e-01,3.830564e-01
min,1.000000e+00,1.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00
25%,2.800000e+02,2.000000e+00,3.727000e+03,4.050000e+02,1.000000e+00,0.000000e+00,0.000000e+00
50%,5.580000e+02,4.000000e+00,5.744000e+03,6.090000e+02,1.000000e+00,0.000000e+00,0.000000e+00
75%,8.380000e+02,6.000000e+00,7.856000e+03,8.370000e+02,1.000000e+00,1.000000e+00,0.000000e+00
max,1.115000e+03,7.000000e+00,4.155100e+04,7.388000e+03,1.000000e+00,1.000000e+00,1.000000e+00


## Data cleaning

### Stores

#### Nulos de CompetitionDistance en df_store

En primera instancia los eliminariamos al ser pocos, pero primero veamos si esa store tiene muchos registros de ventas en train

In [10]:
stores_sin_competitiondistance = df_store[df_store["CompetitionDistance"].isna()]["Store"]

In [11]:
stores_sin_competitiondistance

290    291
621    622
878    879
Name: Store, dtype: int64

Registros de train con los store que no tienen CompetitionDistance

In [12]:
len(df_train[df_train["Store"].isin(stores_sin_competitiondistance)])

2642

Solo son 2642, dentro de más de 1 millón esto son pocos, asi que los eliminaremos

In [13]:
df_store = df_store.dropna(subset=["CompetitionDistance"])

In [14]:
df_store.info()

<class 'pandas.DataFrame'>
Index: 1112 entries, 0 to 1114
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Store                      1112 non-null   int64  
 1   StoreType                  1112 non-null   str    
 2   Assortment                 1112 non-null   str    
 3   CompetitionDistance        1112 non-null   float64
 4   CompetitionOpenSinceMonth  761 non-null    float64
 5   CompetitionOpenSinceYear   761 non-null    float64
 6   Promo2                     1112 non-null   int64  
 7   Promo2SinceWeek            570 non-null    float64
 8   Promo2SinceYear            570 non-null    float64
 9   PromoInterval              570 non-null    str    
dtypes: float64(5), int64(2), str(3)
memory usage: 95.6 KB


#### Nulos en CompetitionOpenSinceMonth y CompetitionOpenSinceYear

Estos nulos son falta de información, por lo que habría que imputar con mediana ya que hacerlo con 0's diría que la competición nunca abrió. Pero si imputamos con mediana lo que va a ocurrir es que utilizaremos datos que irán a "test", por lo que estaríamos creando un pequeño data leakage, así que dejaremos esa imputación para el apartado de feature engineering.

#### Nulos en Promo2SinceWeek y Promo2SinceYear

Estos nulos quieren decir que no hay Promo2 en ese local, asi que lo rellenaremos con 0's

In [16]:
df_store["Promo2SinceWeek"] = df_store["Promo2SinceWeek"].fillna(0)
df_store["Promo2SinceYear"] = df_store["Promo2SinceYear"].fillna(0)

In [17]:
df_store.info()

<class 'pandas.DataFrame'>
Index: 1112 entries, 0 to 1114
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Store                      1112 non-null   int64  
 1   StoreType                  1112 non-null   str    
 2   Assortment                 1112 non-null   str    
 3   CompetitionDistance        1112 non-null   float64
 4   CompetitionOpenSinceMonth  761 non-null    float64
 5   CompetitionOpenSinceYear   761 non-null    float64
 6   Promo2                     1112 non-null   int64  
 7   Promo2SinceWeek            1112 non-null   float64
 8   Promo2SinceYear            1112 non-null   float64
 9   PromoInterval              570 non-null    str    
dtypes: float64(5), int64(2), str(3)
memory usage: 95.6 KB


#### Nulos de PromoInterval

Veamos que categorías tiene el PromoInterval

In [18]:
df_store["PromoInterval"].value_counts()

PromoInterval
Jan,Apr,Jul,Oct     335
Feb,May,Aug,Nov     129
Mar,Jun,Sept,Dec    106
Name: count, dtype: int64

Solo tiene 3, asi que podríamos luego convertirla en dummies sin hacer tan complejo el dataset, completamos con "NoPromo" indicando que no existe promoción

In [19]:
df_store["PromoInterval"] = df_store["PromoInterval"].fillna("NoPromo")

In [20]:
df_store.info()

<class 'pandas.DataFrame'>
Index: 1112 entries, 0 to 1114
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Store                      1112 non-null   int64  
 1   StoreType                  1112 non-null   str    
 2   Assortment                 1112 non-null   str    
 3   CompetitionDistance        1112 non-null   float64
 4   CompetitionOpenSinceMonth  761 non-null    float64
 5   CompetitionOpenSinceYear   761 non-null    float64
 6   Promo2                     1112 non-null   int64  
 7   Promo2SinceWeek            1112 non-null   float64
 8   Promo2SinceYear            1112 non-null   float64
 9   PromoInterval              1112 non-null   str    
dtypes: float64(5), int64(2), str(3)
memory usage: 95.6 KB


#### Hay tiendas duplicadas?

In [30]:
df_store["Store"].duplicated().sum()

np.int64(0)

### Train

#### Convertir Date a tipo datetime

In [22]:
df_train["Date"] = pd.to_datetime(df_train["Date"])

In [23]:
df_train.info()

<class 'pandas.DataFrame'>
RangeIndex: 1017209 entries, 0 to 1017208
Data columns (total 9 columns):
 #   Column         Non-Null Count    Dtype         
---  ------         --------------    -----         
 0   Store          1017209 non-null  int64         
 1   DayOfWeek      1017209 non-null  int64         
 2   Date           1017209 non-null  datetime64[us]
 3   Sales          1017209 non-null  int64         
 4   Customers      1017209 non-null  int64         
 5   Open           1017209 non-null  int64         
 6   Promo          1017209 non-null  int64         
 7   StateHoliday   1017209 non-null  object        
 8   SchoolHoliday  1017209 non-null  int64         
dtypes: datetime64[us](1), int64(7), object(1)
memory usage: 69.8+ MB


#### StateHoliday es object

Veamos porqué. Primero revisar que categorías tiene

In [24]:
df_train["StateHoliday"].unique()

array(['0', 'a', 'b', 'c', 0], dtype=object)

Existen categorías como "0", "a", 0

Entonces sabemos que "0" y 0 significan lo mismo, no hay holiday, asi que convertiremos todo en un string, así el 0, pasa a ser "0"

In [25]:
df_train["StateHoliday"] = df_train["StateHoliday"].astype(str)

In [26]:
df_train.info()

<class 'pandas.DataFrame'>
RangeIndex: 1017209 entries, 0 to 1017208
Data columns (total 9 columns):
 #   Column         Non-Null Count    Dtype         
---  ------         --------------    -----         
 0   Store          1017209 non-null  int64         
 1   DayOfWeek      1017209 non-null  int64         
 2   Date           1017209 non-null  datetime64[us]
 3   Sales          1017209 non-null  int64         
 4   Customers      1017209 non-null  int64         
 5   Open           1017209 non-null  int64         
 6   Promo          1017209 non-null  int64         
 7   StateHoliday   1017209 non-null  str           
 8   SchoolHoliday  1017209 non-null  int64         
dtypes: datetime64[us](1), int64(7), str(1)
memory usage: 69.8 MB


## Guardamos en clean

Ojo que seguimos teniendo nulos en CompetitionOpenSinceMonth y CompetitionOpenSinceYear

In [27]:
df_store.to_csv("../data/clean/store_clean.csv")
df_train.to_csv("../data/clean/train_clean.csv")

## Entendimiento por gráficos

### Merge

Primero combinaremos los store con el train, así podremos obtener solo 1 dataframe y sacar relaciones en base a ello

In [28]:
df = df_train.merge(df_store, on="Store", how="left")

In [29]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1017209 entries, 0 to 1017208
Data columns (total 18 columns):
 #   Column                     Non-Null Count    Dtype         
---  ------                     --------------    -----         
 0   Store                      1017209 non-null  int64         
 1   DayOfWeek                  1017209 non-null  int64         
 2   Date                       1017209 non-null  datetime64[us]
 3   Sales                      1017209 non-null  int64         
 4   Customers                  1017209 non-null  int64         
 5   Open                       1017209 non-null  int64         
 6   Promo                      1017209 non-null  int64         
 7   StateHoliday               1017209 non-null  str           
 8   SchoolHoliday              1017209 non-null  int64         
 9   StoreType                  1014567 non-null  str           
 10  Assortment                 1014567 non-null  str           
 11  CompetitionDistance        1014567 non-null  flo

Aparecieron nulos, esto porque hay tiendas que existen en train, pero no en store, entonces para mejor entendimiento, eliminaremos esos registros

In [31]:
df = df.dropna(subset="StoreType")

In [32]:
df.info()

<class 'pandas.DataFrame'>
Index: 1014567 entries, 0 to 1017208
Data columns (total 18 columns):
 #   Column                     Non-Null Count    Dtype         
---  ------                     --------------    -----         
 0   Store                      1014567 non-null  int64         
 1   DayOfWeek                  1014567 non-null  int64         
 2   Date                       1014567 non-null  datetime64[us]
 3   Sales                      1014567 non-null  int64         
 4   Customers                  1014567 non-null  int64         
 5   Open                       1014567 non-null  int64         
 6   Promo                      1014567 non-null  int64         
 7   StateHoliday               1014567 non-null  str           
 8   SchoolHoliday              1014567 non-null  int64         
 9   StoreType                  1014567 non-null  str           
 10  Assortment                 1014567 non-null  str           
 11  CompetitionDistance        1014567 non-null  float64 

### Gráficos

In [33]:
df.head()

,Store,DayOfWeek,Date,Sales,Customers,Open,Promo,StateHoliday,SchoolHoliday,StoreType,Assortment,CompetitionDistance,CompetitionOpenSinceMonth,CompetitionOpenSinceYear,Promo2,Promo2SinceWeek,Promo2SinceYear,PromoInterval
0,1,5,2015-07-31,5263,555,1,1,0,1,c,a,1270.0,9.0,2008.0,0.0,0.0,0.0,NoPromo
1,2,5,2015-07-31,6064,625,1,1,0,1,a,a,570.0,11.0,2007.0,1.0,13.0,2010.0,"Jan,Apr,Jul,Oct"
2,3,5,2015-07-31,8314,821,1,1,0,1,a,a,14130.0,12.0,2006.0,1.0,14.0,2011.0,"Jan,Apr,Jul,Oct"
3,4,5,2015-07-31,13995,1498,1,1,0,1,c,c,620.0,9.0,2009.0,0.0,0.0,0.0,NoPromo
4,5,5,2015-07-31,4822,559,1,1,0,1,a,a,29910.0,4.0,2015.0,0.0,0.0,0.0,NoPromo
